File: telegram_connector.py
"""
Crystal Bay Travel - Telegram Connector
Интеграция с Telegram Bot API для обработки сообщений
"""

import os
import logging
import requests
from datetime import datetime
from typing import Dict, Optional, List

logger = logging.getLogger(__name__)

class TelegramConnector:
    """Коннектор для работы с Telegram Bot API"""
    
    def __init__(self, token: Optional[str] = None):
        self.token = token or os.environ.get('TELEGRAM_BOT_TOKEN')
        if not self.token:
            raise ValueError("TELEGRAM_BOT_TOKEN не найден")
        
        self.base_url = f"https://api.telegram.org/bot{self.token}"
        self.webhook_url = None
    
    def set_webhook(self, url: str) -> Dict:
        """Установить webhook для получения сообщений"""
        try:
            response = requests.post(
                f"{self.base_url}/setWebhook",
                json={'url': url}
            )
            result = response.json()
            logger.info(f"Webhook установлен: {url}")
            self.webhook_url = url
            return result
        except Exception as e:
            logger.error(f"Ошибка установки webhook: {e}")
            return {'ok': False, 'error': str(e)}
    
    def delete_webhook(self) -> Dict:
        """Удалить webhook"""
        try:
            response = requests.post(f"{self.base_url}/deleteWebhook")
            result = response.json()
            logger.info("Webhook удален")
            self.webhook_url = None
            return result
        except Exception as e:
            logger.error(f"Ошибка удаления webhook: {e}")
            return {'ok': False, 'error': str(e)}
    
    def get_webhook_info(self) -> Dict:
        """Получить информацию о текущем webhook"""
        try:
            response = requests.get(f"{self.base_url}/getWebhookInfo")
            return response.json()
        except Exception as e:
            logger.error(f"Ошибка получения информации о webhook: {e}")
            return {'ok': False, 'error': str(e)}
    
    def send_message(self, chat_id: str, text: str, reply_to_message_id: Optional[str] = None) -> Dict:
        """Отправить текстовое сообщение"""
        try:
            data = {
                'chat_id': chat_id,
                'text': text,
                'parse_mode': 'HTML'
            }
            if reply_to_message_id:
                data['reply_to_message_id'] = reply_to_message_id
            
            response = requests.post(
                f"{self.base_url}/sendMessage",
                json=data
            )
            result = response.json()
            
            if result.get('ok'):
                logger.info(f"Сообщение отправлено в чат {chat_id}")
            else:
                logger.error(f"Ошибка отправки сообщения: {result}")
            
            return result
        except Exception as e:
            logger.error(f"Ошибка отправки сообщения: {e}")
            return {'ok': False, 'error': str(e)}
    
    def send_photo(self, chat_id: str, photo_url: str, caption: Optional[str] = None) -> Dict:
        """Отправить фото"""
        try:
            data = {
                'chat_id': chat_id,
                'photo': photo_url
            }
            if caption:
                data['caption'] = caption
            
            response = requests.post(
                f"{self.base_url}/sendPhoto",
                json=data
            )
            result = response.json()
            
            if result.get('ok'):
                logger.info(f"Фото отправлено в чат {chat_id}")
            else:
                logger.error(f"Ошибка отправки фото: {result}")
            
            return result
        except Exception as e:
            logger.error(f"Ошибка отправки фото: {e}")
            return {'ok': False, 'error': str(e)}
    
    def get_updates(self, offset: Optional[int] = None, limit: int = 100) -> List[Dict]:
        """Получить обновления (polling mode)"""
        try:
            params = {'limit': limit}
            if offset:
                params['offset'] = offset
            
            response = requests.get(
                f"{self.base_url}/getUpdates",
                params=params
            )
            result = response.json()
            
            if result.get('ok'):
                return result.get('result', [])
            else:
                logger.error(f"Ошибка получения обновлений: {result}")
                return []
        except Exception as e:
            logger.error(f"Ошибка получения обновлений: {e}")
            return []
    
    def parse_message(self, update: Dict) -> Optional[Dict]:
        """Парсинг входящего сообщения из webhook/update"""
        try:
            message = update.get('message', {})
            if not message:
                return None
            
            chat = message.get('chat', {})
            from_user = message.get('from', {})
            
            parsed = {
                'message_id': str(message.get('message_id')),
                'chat_id': str(chat.get('id')),
                'from_user_id': str(from_user.get('id')),
                'from_username': from_user.get('username') or from_user.get('first_name', ''),
                'text': message.get('text', ''),
                'message_type': 'text',
                'received_at': datetime.fromtimestamp(message.get('date', 0))
            }
            
            # Определяем тип сообщения
            if message.get('photo'):
                parsed['message_type'] = 'photo'
                photos = message['photo']
                if photos:
                    parsed['media_url'] = photos[-1].get('file_id')
                parsed['text'] = message.get('caption', '')
            
            elif message.get('document'):
                parsed['message_type'] = 'document'
                parsed['media_url'] = message['document'].get('file_id')
                parsed['text'] = message.get('caption', '')
            
            elif message.get('location'):
                parsed['message_type'] = 'location'
                location = message['location']
                parsed['text'] = f"Локация: {location.get('latitude')}, {location.get('longitude')}"
            
            return parsed
        
        except Exception as e:
            logger.error(f"Ошибка парсинга сообщения: {e}")
            return None
    
    def get_bot_info(self) -> Dict:
        """Получить информацию о боте"""
        try:
            response = requests.get(f"{self.base_url}/getMe")
            return response.json()
        except Exception as e:
            logger.error(f"Ошибка получения информации о боте: {e}")
            return {'ok': False, 'error': str(e)}
    
    def mark_chat_read(self, chat_id: str) -> Dict:
        """Отметить сообщения как прочитанные (через sendChatAction)"""
        try:
            response = requests.post(
                f"{self.base_url}/sendChatAction",
                json={
                    'chat_id': chat_id,
                    'action': 'typing'
                }
            )
            return response.json()
        except Exception as e:
            logger.error(f"Ошибка отметки прочитанного: {e}")
            return {'ok': False, 'error': str(e)}

File: integrations/TelegramConnector.js
const BaseConnector = require('./BaseConnector');
const axios = require('axios');
const crypto = require('crypto');

class TelegramConnector extends BaseConnector {
    constructor() {
        super('telegram');
        this.botToken = null;
        this.webhookUrl = null;
        this.apiBase = 'https://api.telegram.org/bot';
    }

    async initialize(config) {
        this.config = config;
        this.botToken = config.bot_token || process.env.TELEGRAM_BOT_TOKEN;
        this.webhookUrl = config.webhook_url || null;

        if (!this.botToken) {
            console.warn('TelegramConnector: Bot token not configured');
            return false;
        }

        return true;
    }

    async connect() {
        if (!this.botToken) {
            throw new Error('Telegram bot token not configured');
        }

        try {
            const response = await axios.get(`${this.apiBase}${this.botToken}/getMe`);
            if (response.data.ok) {
                this.isConnected = true;
                this.botInfo = response.data.result;
                console.log(`TelegramConnector: Connected as @${this.botInfo.username}`);
                return {
                    success: true,
                    bot_info: this.botInfo
                };
            }
        } catch (error) {
            console.error('TelegramConnector: Connection failed:', error.message);
            this.isConnected = false;
            return {
                success: false,
                error: error.message
            };
        }
    }

    async disconnect() {
        this.isConnected = false;
        return { success: true };
    }

    async setWebhook(url) {
        if (!this.botToken) {
            throw new Error('Telegram bot token not configured');
        }

        try {
            const response = await axios.post(`${this.apiBase}${this.botToken}/setWebhook`, {
                url: url,
                allowed_updates: ['message', 'callback_query', 'inline_query']
            });

            if (response.data.ok) {
                this.webhookUrl = url;
                return { success: true, message: 'Webhook set successfully' };
            }
            return { success: false, error: response.data.description };
        } catch (error) {
            return { success: false, error: error.message };
        }
    }

    async deleteWebhook() {
        if (!this.botToken) {
            throw new Error('Telegram bot token not configured');
        }

        try {
            const response = await axios.post(`${this.apiBase}${this.botToken}/deleteWebhook`);
            if (response.data.ok) {
                this.webhookUrl = null;
                return { success: true };
            }
            return { success: false, error: response.data.description };
        } catch (error) {
            return { success: false, error: error.message };
        }
    }

    async sendMessage(chatId, message, options = {}) {
        if (!this.botToken) {
            throw new Error('Telegram bot token not configured');
        }

        try {
            const payload = {
                chat_id: chatId,
                text: message,
                parse_mode: options.parse_mode || 'HTML'
            };

            if (options.reply_markup) {
                payload.reply_markup = options.reply_markup;
            }

            if (options.reply_to_message_id) {
                payload.reply_to_message_id = options.reply_to_message_id;
            }

            const response = await axios.post(`${this.apiBase}${this.botToken}/sendMessage`, payload);

            if (response.data.ok) {
                return {
                    success: true,
                    message_id: response.data.result.message_id,
                    external_message_id: String(response.data.result.message_id),
                    raw_response: response.data.result
                };
            }
            return { success: false, error: response.data.description };
        } catch (error) {
            return { success: false, error: error.message };
        }
    }

    async sendPhoto(chatId, photoUrl, caption = '', options = {}) {
        if (!this.botToken) {
            throw new Error('Telegram bot token not configured');
        }

        try {
            const payload = {
                chat_id: chatId,
                photo: photoUrl,
                caption: caption,
                parse_mode: options.parse_mode || 'HTML'
            };

            const response = await axios.post(`${this.apiBase}${this.botToken}/sendPhoto`, payload);

            if (response.data.ok) {
                return {
                    success: true,
                    message_id: response.data.result.message_id,
                    raw_response: response.data.result
                };
            }
            return { success: false, error: response.data.description };
        } catch (error) {
            return { success: false, error: error.message };
        }
    }

    async sendDocument(chatId, documentUrl, caption = '', options = {}) {
        if (!this.botToken) {
            throw new Error('Telegram bot token not configured');
        }

        try {
            const payload = {
                chat_id: chatId,
                document: documentUrl,
                caption: caption,
                parse_mode: options.parse_mode || 'HTML'
            };

            const response = await axios.post(`${this.apiBase}${this.botToken}/sendDocument`, payload);

            if (response.data.ok) {
                return {
                    success: true,
                    message_id: response.data.result.message_id,
                    raw_response: response.data.result
                };
            }
            return { success: false, error: response.data.description };
        } catch (error) {
            return { success: false, error: error.message };
        }
    }

    async getUpdates(offset = 0, limit = 100) {
        if (!this.botToken) {
            throw new Error('Telegram bot token not configured');
        }

        try {
            const response = await axios.get(`${this.apiBase}${this.botToken}/getUpdates`, {
                params: { offset, limit, timeout: 30 }
            });

            if (response.data.ok) {
                return {
                    success: true,
                    updates: response.data.result.map(u => this.normalizeIncomingMessage(u))
                };
            }
            return { success: false, error: response.data.description };
        } catch (error) {
            return { success: false, error: error.message };
        }
    }

    async handleWebhook(payload, signature = null) {
        try {
            const normalizedMessage = this.normalizeIncomingMessage(payload);
            return {
                success: true,
                message: normalizedMessage,
                raw_payload: payload
            };
        } catch (error) {
            return { success: false, error: error.message };
        }
    }

    normalizeIncomingMessage(update) {
        const message = update.message || update.edited_message || update.channel_post;

        if (!message) {
            return null;
        }

        const sender = message.from || {};
        const chat = message.chat || {};

        let messageType = 'text';
        let content = message.text || '';
        let mediaUrl = null;
        let mediaType = null;

        if (message.photo) {
            messageType = 'photo';
            const photo = message.photo[message.photo.length - 1];
            mediaUrl = photo.file_id;
            mediaType = 'image';
            content = message.caption || '';
        } else if (message.document) {
            messageType = 'document';
            mediaUrl = message.document.file_id;
            mediaType = message.document.mime_type;
            content = message.caption || '';
        } else if (message.voice) {
            messageType = 'voice';
            mediaUrl = message.voice.file_id;
            mediaType = 'audio/ogg';
        } else if (message.video) {
            messageType = 'video';
            mediaUrl = message.video.file_id;
            mediaType = 'video';
            content = message.caption || '';
        } else if (message.sticker) {
            messageType = 'sticker';
            mediaUrl = message.sticker.file_id;
            content = message.sticker.emoji || '🎨';
        } else if (message.location) {
            messageType = 'location';
            content = `📍 ${message.location.latitude}, ${message.location.longitude}`;
        } else if (message.contact) {
            messageType = 'contact';
            content = `👤 ${message.contact.first_name} ${message.contact.last_name || ''}: ${message.contact.phone_number}`;
        }

        return {
            external_message_id: String(message.message_id),
            external_chat_id: String(chat.id),
            channel: 'telegram',
            direction: 'in',
            sender_type: 'customer',
            sender_id: String(sender.id || ''),
            sender_name: `${sender.first_name || ''} ${sender.last_name || ''}`.trim() || sender.username || 'Unknown',
            sender_username: sender.username || null,
            message_type: messageType,
            content: content,
            media_url: mediaUrl,
            media_type: mediaType,
            chat_type: chat.type,
            chat_title: chat.title || null,
            metadata: {
                update_id: update.update_id,
                date: message.date,
                chat_id: chat.id,
                is_bot: sender.is_bot || false
            }
        };
    }

    getStatus() {
        return {
            channel: this.channelName,
            connected: this.isConnected,
            bot_token_set: !!this.botToken,
            webhook_url: this.webhookUrl,
            bot_info: this.botInfo || null
        };
    }
}

module.exports = new TelegramConnector();